### DISCUSSION TITLE
How to use two php versions in same laragon instance?

### DISCUSSION THREAD

**Discussion by anburocky3** [leokhoa/laragon#433] [Q&A]
I have two projects
- Project with 7.4 support
- Project with 8.1 support

I need to run these two project at the same time. For now, we can only switch the entire PHP version. But if i do that, Project 1 won't work. 

How do we solve this issue?

**Comment by Pen-y-Fan** [comment]
Two Laragon versions can be run at the same time. Download the portable version and copy PHP 7.4 and Apache or Nginx into the bin folder. Configure the preferences to use a different port, e.g. 8081. MySQL can still be run from your main installation, with different databases. If you want to keep MySQL separate they would need to run on different ports too.

The PHP 7.4 project will need to go into the www directory of the portable version.

A more complex way is to set a fixed PHP version in Apache:

## Different PHP for Apache and NGINX

It is possible to run Apache with fast CGI fixed to one version of PHP and NGINX running the PHP version set by Laragon:

![0_1666808918363_upload-b27839be-5278-4484-bbc1-9ce665550ea5](https://i.imgur.com/khTv59R.png)

Enable NGINX: Laragon > Menu > Preferences > Services & Ports > tick the two boxes on either side of Nginx and SSL

![0_1666809484087_upload-61ba08c6-85fb-48e6-832a-73ed67d1ff1e](https://i.imgur.com/cU17yVq.png)

Next, stop Laragon and start again, NGINX will be set to ports 8080 & 8443, with whatever PHP version Laragon is configured to use.

## Configure Apache

The fast-CGI modules are already in Laragon, under laragon/etc/apache2/modules/mod_fcgid-2.3.10-win64-VS16.so, this can be set to override the mod_php.conf file automatically generated by Laragon.

Create a file **fcgid.conf** as follows (change E: to your drive letter) and place it in E:\laragon\etc\apache2\sites-enabled. Note forward slashes are used as directory separators in the .conf file.


```conf
# source info: https://www.php.net/manual/en/install.windows.apache2.php
# E:\laragon\bin\apache\httpd-2.4.54-win64-VS16\modules
LoadModule fcgid_module E:/laragon/etc/apache2/modules/mod_fcgid-2.3.10-win64-VS16.so

# Where is your php.ini file?
FcgidInitialEnv PHPRC "E:/laragon/bin/php/php-8.1.10-nts-Win32-vs16-x64"
FcgidWrapper "E:/laragon/bin/php/php-8.1.10-nts-Win32-vs16-x64/php-cgi.exe" .php

<FilesMatch \.php$>
    SetHandler fcgid-script
</FilesMatch>
```

⚠️ Note: The PHP version needs to be non-thread safe (NTS) for fast-CGI to be automatically selected by Laragon.

Stop Laragon and Start it again. Apache will be fixed to PHP 8.1.10 (or whatever your set in the fcgid.conf above). Once you want to revert Apache back to Laragon's setting rename the **fcgid.conf** file **fcgid.conf.tmp**. Stop Laragon and start again to remove the config. If you need to fix the version in the future remove the .tmp ending to revert it to a .conf file.

I hope this helps.

**Comment by anburocky3** [comment]
@Pen-y-Fan In Portable version, the apache service is grayed out.
![image](https://user-images.githubusercontent.com/15264938/224686192-f4c91ed7-660e-4eac-9c19-aa654d8e5cde.png)

I copied my folder from `laragon6` (Full version) to `laragon6-portable` (Portable version). 

I can't access the localhost while running `laragon6-portable`
![image](https://user-images.githubusercontent.com/15264938/224686528-f4173fa6-85f9-43b1-840e-1ac4ddb3b494.png)

**Reply by Pen-y-Fan** [reply]
What port have you set for the portable version? Try http://localhost:8081

![laragon portible](https://user-images.githubusercontent.com/40126936/224795778-ed4eee3f-4f87-453e-940d-f72356bd756e.png)

**Reply by Pen-y-Fan** [reply]
![incorrect link](https://user-images.githubusercontent.com/40126936/224799216-571464f6-4c83-47ca-bf0d-c74e7c4ec78b.png)

![correct link](https://user-images.githubusercontent.com/40126936/224800041-ba4da26a-718a-4e94-b99d-d2333fb4e1cd.png)

**Comment by Pen-y-Fan** [comment]
This took a little more setup to get it working:

![image](https://user-images.githubusercontent.com/40126936/224854126-b22a5ae9-4f0d-4aae-b8a0-9552412b1b1c.png)

The full version needs to be different setup from the portable otherwise when changes are detected by Laragon the components (e.g. Apache) will be restarted. 

- The full version can be setup with PHP 8.1, Apache and MySQL, with a domain of test.
- The portable version with PHP 7.4, Nginx, Maria DB* and a domain of local.
- The host file needs to be manually setup for each project:

```text
127.0.0.1      monitor.test
127.0.0.1      calendar.local       #laragon magic!
```

Only one of the projects can be registered with laragon.crt for SSL. The other project will allow https, but the browser will complain. Once the "risk is accepted" the project will render Ok.

I hope this helps.

\* the MySQL DB server could be used for both projects, just with a different database.
